File: c++/p001-p099/leetcode005.cpp
// <!-- encoding UTF-8 --!>
/*****************************************************************************
*                     ----Stay Hungry Stay Foolish----                       *
*   @author	:   Shen                                                         *
*   @name   :   Leetcode 005 Longest Palindromic Substring                   *
*****************************************************************************/

#include <bits/stdc++.h>
using namespace std;
typedef long long int64;
template<class T>inline bool updateMin(T& a, T b){ return a > b ? a = b, 1: 0; }
template<class T>inline bool updateMax(T& a, T b){ return a < b ? a = b, 1: 0; }

class Solution {
public:
	static const int MAXL = 1024;
	char Ma[MAXL * 2];
	int Mp[MAXL * 2];
	string longestPalindrome(string s)
	{
		int sz = s.size();
		int l = 0;
		/// manacher algorithm O(n)
		Ma[l++] = '$'; Ma[l++] = '#';
		for (int i = 0; i < sz; i++)
		{
			Ma[l++] = s[i];
			Ma[l++] = '#';
		}
		Ma[l] = 0;
		int mx = 0, id = 0;
		for (int i = 0; i < l; i++)
		{
			Mp[i] = mx > i ? min(Mp[2 * id - i], mx - i) : 1;
			while (Ma[i + Mp[i]] == Ma[i - Mp[i]]) Mp[i]++;
			if (i + Mp[i] > mx)
			{
				mx = i + Mp[i];
				id = i;
			}
		}
		/// get palindrome
		int pos = std::max_element(Mp, Mp + l) - Mp;
		int val = Mp[pos] - 1;
		string ret = "";
		for (int start = pos - val, end = pos + val; start <= end; start++)
			if (Ma[start] == '#' || Ma[start] == '$')
				continue;
			else ret += Ma[start];
		return ret;
	}
};

void test()
{

}

int main()
{

	return 0;
}

File: ruby/leetcode005.rb
# @param {String} s
# @return {String}
# 308 ms
def longest_palindrome(s)
    # manacher algorithm O(n)
    s = '$' + s;
    ma = s.each_char.map { |e| e + '#' }.inject(:+)
    l = ma.length
    mp = Array.new(l) { 0 }
    mx, id = 0, 0
    l.times { |i|
        mp[i] = mx > i ? [mp[2 * id - i], mx - i].min : 1
        while i - mp[i] > 0 && ma[i + mp[i]] == ma[i - mp[i]] do 
            mp[i] = mp[i] + 1
        end
        mx, id = i + mp[i], i if i + mp[i] > mx
    }
    # get palindrome
    pkg = mp.zip(0..l).max { |a, b| a[0] <=> b[0] }
    pos, val = pkg[1], pkg[0] - 1
    ma[pos - val .. pos + val].delete '#'.delete '$'
end